lib/compiler/src/beam_ssa.erl

```diff
@@ -651,12 +651,18 @@ is_commutative('=/=') -> true;
 is_commutative('/=') -> true;
 is_commutative(_) -> false.
 
-def_used_1([#b_blk{is=Is,last=Last}|Bs], Preds, Def0, Used0) ->
-    {Def,Used1} = def_used_is(Is, Preds, Def0, Used0),
-    Used = ordsets:union(used(Last), Used1),
-    def_used_1(Bs, Preds, Def, Used);
-def_used_1([], _Preds, Def, Used) ->
-    {ordsets:from_list(Def),Used}.
+def_used_1([#b_blk{is=Is,last=Last}|Bs], Preds, Def0, UsedAcc) ->
+    {Def,Used} = def_used_is(Is, Preds, Def0, used(Last)),
+    case Used of
+        [] ->
+            def_used_1(Bs, Preds, Def, UsedAcc);
+        [_|_] ->
+            def_used_1(Bs, Preds, Def, [Used|UsedAcc])
+    end;
+def_used_1([], _Preds, Def0, UsedAcc) ->
+    Def = ordsets:from_list(Def0),
+    Used = umerge(UsedAcc),
```

Comment:
Curiosity: is the sorting necessary for the algorithm to work or are we using ordsets for speed under certain operations?

Reply:
Well, both, really. The algorithm manipulates sets and in this case `ordsets` is the most efficient representation.

The caller will do several set operations that require sets represented as sorted lists:

https://github.com/erlang/otp/blob/5149af57dc17afae1ea97fb939192dd9ccb71d82/lib/compiler/src/beam_ssa_pre_codegen.erl#L2136-L2140

When doing union, intersection, and subtraction operations on sets, the `ordsets` representation is usually the fastest set representation.

Depending on the mix of operations on sets, the compiler tries to use the most efficient representation. In other parts of the compiler, `cerl_sets` (sets based on maps) or `gb_sets` are used.

Reply:
Thanks @bjorng!

I have noticed that, most of the time, the variables we are working with in that code snippet are integers, which are likely the ones generated by the compiler. Therefore I wondered what if we can optimize this part of the problem into a bitset?

For fun, I have implemented an intset in Elixir here: https://github.com/josevalim/intset. In the benchmark folder, you will find comparisons between union/intersection/subtract between ordsets and intset. The results are also in the benchmark files for convenience. In a nutshell, for above 10 entries, intset is faster than ordsets. For above 1000 entries, it is an order of magnitude faster. It also uses considerably less memory. The only operation slower in intsets is `from_list`, which is 2x-4x slower, regardless of size. I haven't implemented traversal (traversals would definitely be expensive though). 

My idea would be to have a special kind of "varset" in `beam_ssa` that stores numbered variables in the intset and the others in the ordset. The only issue I found so far is that sometimes, those numbered variables are named `{@ssa_return, number}` (or `{@ssa_bool, number}`).

Another point is the size of the integers. We need to work with small integers and the safest choice in all systems is 28 bits - which I currently use. We can also change the number of bits based on the wordsize. I can do some benchmarking around this if required.

Of course, there may be faster implementations of this approach. Maybe using binaries (intersection/union/subtract BIFs could be added to binaries) or HiPE bit arrays (although I guess we don't want to rely on that here). For now, what I would love some input on:

1. Do you think it is worth to continue pursuing "varsets"? Can you see "varsets"  being useful in other areas/passes? 

2. Are `@ssa_return` and `@ssa_bool` annotations? If so, is it required for them to be part of the name or can we store them separately? Finally, once a variable is annotated, does it ever change its annotation? I would think it doesn't, since currently the annotation is part of the name.

3. Finally, are only variables stored in those sets? So far those sets have only `{b_var, _}` tuples, but since the whole tuple is stored, I wonder if they are other cases.

Thank you!

Reply:
Intsets are looking great in theory but putting them in practice is much harder! :sweat_smile: In any case, I found another approach that is slightly faster and should use less memory: https://github.com/erlang/otp/pull/2314